core/src/main/java/com/linecorp/armeria/internal/common/PercentEncoder.java

```diff
@@ -0,0 +1,191 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2008 The Guava Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linecorp.armeria.internal.common;
+
+import io.netty.util.internal.StringUtil;
+
+public final class PercentEncoder {
+
+    // Forked from netty-4.1.43 and guava-28.1
+    // https://github.com/netty/netty/blob/bd8cea644a07890f5bada18ddff0a849b58cd861/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java
+    // https://github.com/google/guava/blob/13e39cd167a49aad525be462e61d9e5f2b1781ec/guava/src/com/google/common/net/PercentEscaper.java
+
+    private static final char[] UTF_UNKNOWN = { '%', '3', 'F' }; // Percent encoded question mark
+    private static final char[] UPPER_HEX_DIGITS = "0123456789ABCDEF".toCharArray();
+    private static final byte[] SAFE_OCTETS = new byte[Character.MAX_VALUE + 1];
+
+    static {
+        // Unreserved characters with '*' because most browsers such as Chrome and Firefox do not encode '*'.
+        // See https://tools.ietf.org/html/rfc3986#section-2.3
+        final String safeOctetStr = "-_.~*abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+        for (int i = 0; i < safeOctetStr.length(); i++) {
+            SAFE_OCTETS[safeOctetStr.charAt(i)] = -1;
+        }
+    }
+
+    /**
+     * Encodes the specified string using
+     * <a href="https://tools.ietf.org/html/rfc3986#section-2.1">Percent-Encoding</a> and appends it to the
+     * specified {@link StringBuilder}.
+     */
+    public static void encodeComponent(StringBuilder buf, String s) {
```

Comment:
Use TemporaryThreadLocals for consistancy?

Reply:
It can't be because the callers mush pass `StringBuilder`.

Reply:
Ah, I see.